spring-pulsar/src/main/java/org/springframework/pulsar/listener/DefaultPulsarMessageListenerContainer.java

```diff
@@ -453,27 +461,21 @@ private void invokeRecordListenerErrorHandler(AtomicBoolean inRetryMode, Message
 				inRetryMode.set(true);
 			}
 			else {
-				if (inRetryMode.get()) {
-					inRetryMode.set(false);
-				}
+				inRetryMode.compareAndSet(true, false);
 				// retries exhausted - recover the message
 				this.pulsarConsumerErrorHandler.recoverMessage(this.consumer, message, e);
 				// retries exhausted - if record ackmode, acknowledge, otherwise normal
 				// batch ack at the end
-				if (this.containerProperties.getAckMode() == AckMode.RECORD) {
+				if (this.isRecordAck) {
 					handleAck(message);
 				}
 			}
 		}
 
 		private void pendingMessagesHandledSuccessfully(AtomicBoolean inRetryMode,
 				AtomicBoolean messagesPendingInBatch) {
-			if (inRetryMode.get()) {
-				inRetryMode.set(false);
-			}
-			if (messagesPendingInBatch.get()) {
-				messagesPendingInBatch.set(false);
-			}
+			inRetryMode.compareAndSet(true, false);
```

Comment:
This is just an implementation detail, so definitely no biggie. But on a second look, seems this code is not executed concurrently in any point, only sequentially inside the main `while` and `Message` loops. So unless I'm missing something a simple `boolean` might be simpler and avoid confusion.

Reply:
The reason why we went with `AtomicBoolean` is so that we can modify it inside some lambda expressions. If it is primitive `boolean` we need to make it `final` to use inside the lambda, but then we cannot modify it. `AtomicBoolean` fixes that issue. You are right; this code is not accessed concurrently from anywhere else.